### PR TITLE
Implemented fully BMI compliant transport code into troute

### PR DESCRIPTION
**Pull request by JurgenZach-NOAA**
Conversion of data frames in the DA forcing module into BMI compliant arrays and reassembly in troute: wrote routines for all dataframe import into troute, and applied it to most dataframes

## Additions

- Modules with auxiliary functions to help convert dataframes to/from arrays:

-- bmi_df2array: Functions to disassemble dataframes into constituent numpy arrays (or individual variables) of either integer or float types. This includes converting any string-based dataframe entries (such as gauge IDs) to arrays based on ASCII-encoding, wherein one string array-like object gets converted into two 1D numpy arrays: a series of ASCIIs in one int array and an array comprised of the length of each string entry. Datetime entries are converted to seconds relative to a troute-internal time reference (which is, a priori, arbitrary). 

-- bmi_array2df: The reverse of the latter. If applicable (for string-like entries such as gauge IDs), first reassembling a list of strings from the two constituent files (ASCII encoded array and ASCII string length array); or, for time arrays, reassembling lists of Datetime objects. Then, reassemble the applicable dataframes. 

## Removals

- On the t-route side, the dataframe import option is kept as a debugging feature, which can be enables through the flag "legacy_bmi_df" in DataAssimilation. This shall be removed in the near future. 

## Changes

- Lion's share of the code changes are in model_DAforcing (dataframe -> array conversions), and DataAssimilation (array conversions -> dataframe. The latter is implemented in the {NudgingDA, PersistenceDA, RFCDA} child classes of AbstractDA. 

- In the (dataframe -> array) direction, the following dataframes were converted: usgs_df, reservoir_usgs_df, reservoir_usace_df, reservoir_rfc_dataframe, q0/t0, waterbody_df, lastobs

- In the (array -> dataframe) direction, the following dataframes were implemented for bmi import into troute: usgs_df, reservoir_usgs_df, reservoir_usace_df, reservoir_rfc_dataframe. There was existing code for the last three (lite_restart and lastobs), which shall be harmonized with the current commit in the near future. 

## Testing

1. Should not interfere with normal operation. A full-featured BMI test case will be constructed. 

## Notes

-

## Todos

- Harmonizing lite_restart and lastobs code with the existing additions. 

## Checklist

- [ ] PR has an informative and human-readable title
- [ ] Changes are limited to a single goal (no scope creep)
- [ ] Code can be automatically merged (no conflicts)
- [ ] Code follows project standards (link if applicable)
- [ ] Passes all existing automated tests
- [ ] Any _change_ in functionality is tested
- [ ] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Placeholder code is flagged / future todos are captured in comments
- [ ] Visually tested in supported browsers and devices (see checklist below :point_down:)
- [ ] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
- [ ] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

### Target Environment support

- [ ] Windows
- [ ] Linux
- [ ] Browser

### Accessibility

- [ ] Keyboard friendly
- [ ] Screen reader friendly

### Other

- [ ] Is useable without CSS
- [ ] Is useable without JS
- [ ] Flexible from small to large screens
- [ ] No linting errors or warnings
- [ ] JavaScript tests are passing

### CODE REVIEW

**Comment by JurgenZach-NOAA** [comment]
Checks have passed now, adapted location of modules in folder structure

File: src/troute-network/troute/DataAssimilation.py [hunk 1]
```diff
@@ -62,16 +94,57 @@ def __init__(self, network, from_files, value_dict, da_run=[]):
...
+                        # Decode station ID axis
+                        stationAxisName = 'stationId'
+                        usgs_df = a2df._stations_retrieve_from_arrays(\
+                                df_withDates_usgs, stationArray_usgs, \
+                                stationStringLengthArray_usgs, stationAxisName)
+                
+                    #usgs_df = usgs_df.join(network.link_gage_df.reset_index().set_index('gages'),how='inner').set_index('link').sort_index()
+                    usgs_df = network.link_gage_df.reset_index().set_index('gages').join(usgs_df).set_index('link').sort_index()
+
+                    self._usgs_df = _reindex_link_to_lake_id(usgs_df, network.link_lake_crosswalk)
```
Comment (shorvath-noaa) [inline src/troute-network/troute/DataAssimilation.py:143]: Should these be un-indented 1 level? So they run regardless of `legacy_bmi_df`?

File: src/troute-network/troute/DataAssimilation.py [hunk 2]
```diff
@@ -202,55 +276,134 @@ def __init__(self, network, from_files, value_dict, da_run=[]):
...
+                        reservoir_usgs_param_df = pd.DataFrame(
+                            data = 0, 
+                            index = reservoir_usgs_df.index ,
+                            columns = ['update_time']
+                        )
+                        reservoir_usgs_param_df['prev_persisted_outflow'] = np.nan
+                        reservoir_usgs_param_df['persistence_update_time'] = 0
+                        reservoir_usgs_param_df['persistence_index'] = 0
+                    else:
+                        reservoir_usgs_param_df = pd.DataFrame()
```
Comment (shorvath-noaa) [inline src/troute-network/troute/DataAssimilation.py:339]: Should these be un-indented 1 level? So they run regardless of legacy_bmi_df?

File: src/troute-network/troute/DataAssimilation.py [hunk 3]
```diff
@@ -1428,90 +1629,3 @@ def _rfc_timeseries_qcqa(discharge,stationId,synthetic,totalCounts,timestamp,tim
...
-    reservoir_rfc_param_df['use_rfc'].fillna(False, inplace=True)
-    reservoir_rfc_param_df['totalCounts'].fillna(0, inplace=True)
-    reservoir_rfc_param_df['da_timestep'].fillna(0, inplace=True)
-    # Make sure columns are the correct types
-    reservoir_rfc_param_df['totalCounts'] = reservoir_rfc_param_df['totalCounts'].astype(int)
-    reservoir_rfc_param_df['da_timestep'] = reservoir_rfc_param_df['da_timestep'].astype(int)
-    reservoir_rfc_param_df['update_time'] = 0
-    reservoir_rfc_param_df['rfc_persist_days'] = rfc_parameters.get('reservoir_rfc_forecast_persist_days', 11)
-
-    return reservoir_rfc_df, reservoir_rfc_param_df
```
[the remark this answers is not in the document]
Comment (JurgenZach-NOAA) [reply src/troute-network/troute/DataAssimilation.py:1517]: What do you mean: rebase on master?

File: src/bmi_DAforcing.py [hunk 4]
```diff
@@ -59,6 +59,85 @@ def __init__(self):
...
+        'datesSecondsArray_reservoir_usace',
+        'nDates_reservoir_usace',
+        'stationArray_reservoir_usace',
+        'stationStringLengthArray_reservoir_usace',
+        'nStations_reservoir_usace',
+        'reservoir_usace_Array'
+    ]
+
+    # RFC DF
+    _BMI_output_reservoir_usgs = [
```
Comment (kumdonoaa) [inline src/bmi_DAforcing.py:96]: Isn't this rfc so _rfc instead of _usgs?

File: src/troute-network/troute/DataAssimilation.py (hunk 1 above)
Comment (JurgenZach-NOAA) [reply src/troute-network/troute/DataAssimilation.py:143]: Yes, agreed!

File: src/troute-network/troute/DataAssimilation.py [hunk 5]
```diff
@@ -517,58 +670,106 @@ class RFCDA(AbstractDA):
...
+                    self._reservoir_rfc_param_df['timeseries_idx'] = new_timeseries_idx
+                    # Fill in NaNs with default values.
+                    self._reservoir_rfc_param_df['use_rfc'].fillna(False, inplace=True)
+                    self._reservoir_rfc_param_df['totalCounts'].fillna(0, inplace=True)
+                    self._reservoir_rfc_param_df['da_timestep'].fillna(0, inplace=True)
+                    # Make sure columns are the correct types
+                    self._reservoir_rfc_param_df['totalCounts'] = self._reservoir_rfc_param_df['totalCounts'].astype(int)
+                    self._reservoir_rfc_param_df['da_timestep'] = self._reservoir_rfc_param_df['da_timestep'].astype(int)
+                    self._reservoir_rfc_param_df['update_time'] = 0
+                    self._reservoir_rfc_param_df['rfc_persist_days'] = rfc_parameters.get('reservoir_rfc_forecast_persist_days', 11)
```
[the remark this answers is not in the document]
Comment (JurgenZach-NOAA) [reply src/troute-network/troute/DataAssimilation.py:762]: Fixed

File: src/troute-network/troute/DataAssimilation.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (JurgenZach-NOAA) [reply src/troute-network/troute/DataAssimilation.py]: Fixed.

File: src/bmi_DAforcing.py (hunk 4 above)
Comment (JurgenZach-NOAA) [reply src/bmi_DAforcing.py:96]: Fixed it,, thanks!

**Review by shorvath-noaa — APPROVED** [review]

**Pull request merged**